### PR TITLE
Anuj/444/documenting observability

### LINKED ISSUE #444: Document Observability aspect of the solution
**DoD**:
- The following is documented:
  - What Telemetry is recorded and how it is recorded
    - Synapse pipeline run metrics (both from within Synapse workspace and from Log Analytics)
    - Custom metrics of the pipeline run sent to AppInsights via OpenCensus library
  - Where to look at them
    - Screenshots of the telemetry from Log Analytics, Synapse workspace, AppInsights
  - Example of an insight that can be viewed
- These are documented here in this section of the main README: https://github.com/Azure-Samples/modern-data-warehouse-dataops/blob/e2e/parking-sensors-synapse/e2e_samples/parking_sensors_synapse/README.md#observability--monitoring

### PR DESCRIPTION
**Pull request by promisinganuj**
# Type of PR
- Documentation changes
- Code changes

## Purpose
To document the observability in the "parking-sensor" synapse version. It includes:

- Checking pipeline/trigger/activity run status from synapse studio and log analytics workspace.
- Illustrating how custom logs are written to log analytics workspace using "log4j" logger.
- Illustrating how custom logs/metrics are sent to app insights using "opencensus" library.
- Relevant screenshots.
- Documentation update.

Here is a summary of changes:
- Enabled diagnostic settings for Synapse as part of infrastructure deployment
```text
/infrastructure/modules/diagnostic_settings.bicep (A)
/infrastructure/main.bicep (M)
```

- Fixed an existing code issue so that the deployment can continue
```text
/scripts/deploy_synapse_artifacts.sh (M)
```

- Relevant code changes for consistency
```text
/synapse/notebook/02_standardize.ipynb (M)
/synapse/notebook/03_transform.ipynb (M)
```

- Documentation update
```text
/docs/observability.md (A)
/README.md (M)
```

- Screenshots added
```text
/docs/images/observability-app-insights-logs-in-monitor.png (A)
/docs/images/observability-app-insights-logs-in-synapse.png (A)
/docs/images/observability-log4j-logs-in-monitor.png (A)
/docs/images/observability-log4j-logs-in-synapse.png (A)
/docs/images/observability-pipeline-runs-in-monitor.png (A)
/docs/images/observability-pipeline-runs-in-synapse.png (A)
```

## Does this introduce a breaking change? If yes, details on what can break
No

## Author pre-publish checklist

- [ ] Added test to prove my fix is effective or new feature works
- [x] No PII in logs
- [x] Made corresponding changes to the documentation

## Validation steps
N/A

## Issues Closed or Referenced

- Closes #444

### CODE REVIEW

File: e2e_samples/parking_sensors_synapse/docs/observability.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+```python
+# Getting Application Insights instrumentation key
+appi_key = token_library.getSecretWithLS(keyvaultlsname,"applicationInsightsKey")
+
+# Enable App Insights
+aiLogger = logging.getLogger(__name__)
+aiLogger.addHandler(AzureEventHandler(connection_string = 'InstrumentationKey=' + appi_key))
+```
+
+Note that it's also possible to set "custom_dimentions" with this approach which can be passed to the logger as a JSON object:
```
Comment (devlace) [inline e2e_samples/parking_sensors_synapse/docs/observability.md:43]: nit: typo, should be custom_dimensions

File: e2e_samples/parking_sensors_synapse/docs/observability.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+```python
+properties = {"custom_dimensions": {"pipeline": pipelinename, "run_id": loadid, "parkingbay_count": parkingbay_count, "sensordata_count": sensordata_count, "parkingbay_malformed_count": parkingbay_malformed_count, "sensordata_malformed_count": sensordata_malformed_count}}
+aiLogger.info("Standardize (ai): Completed at " + datetime.now().strftime("%Y-%m-%d %H:%M:%S"), extra=properties)
+```
+
+Once the pipeline "P_Ingest_MelbParkingData" has been executed, this logging information can be checked in Synapse Studio as shown below:
+
+![observability-app-insights-logs-in-synapse](images/observability-app-insights-logs-in-synapse.png)
+
+The same information can also be queried in Azure Monitor. For that, the relevant application insights instance has been selected for the scope under Azure Monitor -> Logs. The query is based on "customEvents" logs. The "customDimentions" has been parsed to get the individual metrics.
```
Comment (devlace) [inline e2e_samples/parking_sensors_synapse/docs/observability.md:54]: also should be customerDimensions

File: e2e_samples/parking_sensors_synapse/docs/observability.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    env = mssparkutils.env
+    formatted_msg = f'Standardize (log4j): {msg}~{pipelinename}~{env.getJobId()}~{env.getPoolName()}~{env.getWorkspaceName()}~{env.getUserId()}'
+    logger.info(formatted_msg)
+```
+
+Within the Synapse Studio, these log messages can be found under Monitor -> Apache Spark applications -> Logs -> Driver (stderr) as shown below:
+
+![observability-log4j-logs-in-synapse](images/observability-log4j-logs-in-synapse.png)
+
+For querting this information in Azure Monitor, go to Logs and select the relevant log analytics workspace instance as the scope. The custom logs get logged under Custom Logs -> SparkLoggingEvent_CL, here is a screenshot:
```
Comment (devlace) [inline e2e_samples/parking_sensors_synapse/docs/observability.md:95]: nit: querying*

File: e2e_samples/parking_sensors_synapse/synapse/notebook/02_standardize.ipynb [hunk 4]
```diff
@@ -366,25 +186,29 @@
...
         "from datetime import datetime\r\n",
         "\r\n",
+        "# Getting Application Insights instrumentation key\r\n",
+        "appi_key = token_library.getSecretWithLS(keyvaultlsname,\"applicationInsightsKey\")\r\n",
+        "\r\n",
         "# Enable App Insights\r\n",
-        "aiLogger = logging.getLogger(\"ParkingSensorLogs-Standardize\")\r\n",
+        "aiLogger = logging.getLogger(__name__)\r\n",
         "aiLogger.addHandler(AzureEventHandler(connection_string = 'InstrumentationKey=' + appi_key))\r\n",
         "#logger.addHandler(AzureLogHandler(connection_string = 'InstrumentationKey=' + appi_key))\r\n",
```
Comment (devlace) [inline e2e_samples/parking_sensors_synapse/synapse/notebook/02_standardize.ipynb:197]: Minor cleanup, Perhaps we can remove this commented out line?

**Review by devlace — APPROVED** [review]
Just some minor nitpicks, but otherwise, looks good! Thanks for cleanin gup the Observability section in the notebooks as well :)

File: e2e_samples/parking_sensors_synapse/docs/observability.md (hunk 1 above)
Comment (promisinganuj) [reply e2e_samples/parking_sensors_synapse/docs/observability.md]: Fixed.

File: e2e_samples/parking_sensors_synapse/docs/observability.md (hunk 2 above)
Comment (promisinganuj) [reply e2e_samples/parking_sensors_synapse/docs/observability.md]: Fixed.

File: e2e_samples/parking_sensors_synapse/docs/observability.md (hunk 3 above)
Comment (promisinganuj) [reply e2e_samples/parking_sensors_synapse/docs/observability.md]: Fixed.

File: e2e_samples/parking_sensors_synapse/synapse/notebook/02_standardize.ipynb (hunk 4 above)
Comment (promisinganuj) [reply e2e_samples/parking_sensors_synapse/synapse/notebook/02_standardize.ipynb]: Removed the reference to "AzureLogHandler" altogether (from both the notebooks) and added the actual queries as used in the screenshots.

**Pull request merged**